File: tripod_hr/tripod_hr/page/hr_budget_dashboard/hr_budget_dashboard.py
import frappe

REGION_ORDER = {"UAE": 1, "KSA": 2, "": 9}
UNIT_ORDER = {
    "Fit Out UAE": 1, "Dubai Production": 2, "Dubai Office": 3,
    "KSA Office": 4, "KSA National": 5, "KSA Production": 6, "KSA Fit Out": 7,
    "Logistics": 8, "Admin": 9, "Tap Gulf": 10,
}
UNIT_REGION = {
    "Fit Out UAE": "UAE", "Dubai Production": "UAE", "Dubai Office": "UAE",
    "KSA Office": "KSA", "KSA National": "KSA", "KSA Production": "KSA", "KSA Fit Out": "KSA",
    "Logistics": "KSA", "Admin": "KSA", "Tap Gulf": "KSA",
}
OWNERS = ["TM-EMP-0021", "TM-EMP-0022"]        # excluded from budget entirely
CEO = "TGK-EMP-0284"                            # cost split across two offices
CEO_DUBAI_SHARE = 0.40                          # 40% Dubai Office, 60% stays KSA Office

CEO_ALLOC_LABEL = "CEO cost allocation (40%)"
MONTH_NAMES = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
               "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

MONTHS = ["2026-07", "2026-08", "2026-09", "2026-10", "2026-11", "2026-12"]
MONTH_LBL = ["Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


@frappe.whitelist()
def get_budget_dashboard(company=None):
    company = company or None
    active = _active(company)
    plan = _plan(company)
    ramp = _ramp(company, active["current_salary"])
    designations = _designations(company)
    movement = _movement(company, active["current_ctc"])
    return {
        "active": active,
        "plan": plan,
        "ramp": ramp,
        "designations": designations,
        "movement": movement,
    }


def _cond(company, alias=""):
    p = (alias + ".") if alias else ""
    cond = f"{p}status = 'Active'"
    vals = {}
    if company:
        cond += f" AND {p}company = %(company)s"
        vals["company"] = company
    return cond, vals


def _active(company):
    cond, vals = _cond(company)
    rows = frappe.db.sql(
        f"""
        SELECT
            COALESCE(NULLIF(custom_budget_unit, ''), '(untagged)') AS unit,
            COALESCE(custom_region, '')                            AS region,
            COUNT(*)                                               AS hc,
            SUM(COALESCE(custom_total_salary, 0))                  AS salary,
            SUM(COALESCE(custom_monthly_ctc, 0))                   AS ctc,
            SUM(COALESCE(custom_annual_ctc, 0))                    AS ctc_yr
        FROM `tabEmployee`
        WHERE {cond}
          AND name NOT IN %(owners)s
        GROUP BY unit, region
        """,
        dict(vals, owners=tuple(OWNERS)), as_dict=True,
    )

    by_unit = {}
    untagged = {"hc": 0, "salary": 0, "ctc": 0, "ctc_yr": 0}
    for r in rows:
        if r["unit"] == "(untagged)":
            for k in untagged:
                untagged[k] += r[k] or 0
            continue
        u = by_unit.setdefault(r["unit"], {"hc": 0, "salary": 0, "ctc": 0, "ctc_yr": 0})
        for k in u:
            u[k] += r[k] or 0

    if not company:
        ceo = frappe.db.get_value(
            "Employee", CEO,
            ["custom_budget_unit", "custom_total_salary", "custom_monthly_ctc", "custom_annual_ctc"],
            as_dict=True,
        )
        if ceo and ceo.custom_budget_unit in by_unit:
            share = CEO_DUBAI_SHARE
            home = by_unit[ceo.custom_budget_unit]
            dubai = by_unit.setdefault("Dubai Office", {"hc": 0, "salary": 0, "ctc": 0, "ctc_yr": 0})
            for fld, key in [("custom_total_salary", "salary"),
                             ("custom_monthly_ctc", "ctc"),
                             ("custom_annual_ctc", "ctc_yr")]:
                amt = (ceo.get(fld) or 0) * share
                home[key] -= amt
                dubai[key] += amt
            # headcount stays in CEO's home unit; only cost is split

    regions = {}
    for unit, m in by_unit.items():
        reg = UNIT_REGION.get(unit, "")
        regions.setdefault(reg, []).append((unit, m))

    out = []
    grand = {"hc": 0, "salary": 0, "ctc": 0, "ctc_yr": 0}
    for reg in sorted(regions, key=lambda x: REGION_ORDER.get(x, 9)):
        sub = {"hc": 0, "salary": 0, "ctc": 0, "ctc_yr": 0}
        unit_rows = []
        for unit, m in sorted(regions[reg], key=lambda t: UNIT_ORDER.get(t[0], 9)):
            unit_rows.append(_row(reg, unit, m, "unit"))
            for k in sub:
                sub[k] += m[k]
        out.extend(unit_rows)
        out.append(_row(reg, reg + " subtotal", sub, "subtotal"))
        for k in grand:
            grand[k] += sub[k]

    if untagged["hc"]:
        out.append(_row("", "Untagged (to tag)", untagged, "untagged"))
        for k in grand:
            grand[k] += untagged[k]

    out.append(_row("", "TOTAL", grand, "grand"))
    return {"rows": out, "current_ctc": grand["ctc"], "current_salary": grand["salary"]}


def _row(region, unit, m, kind):
    return {
        "region": region, "unit": unit, "kind": kind,
        "hc": int(m["hc"]),
        "salary": round(m["salary"]),
        "salary_yr": round(m["salary"] * 12),
        "ctc": round(m["ctc"]),
        "ctc_yr": round(m.get("ctc_yr", m["ctc"] * 12)),
    }


def _plan(company):
    cond = "status = 'Open'"
    vals = {}
    if company:
        cond += " AND company = %(company)s"
        vals["company"] = company
    rows = frappe.db.sql(
        f"""
        SELECT budget_unit AS unit,
               COUNT(*) AS hc,
               SUM(COALESCE(monthly_salary, 0)) AS salary,
               SUM(COALESCE(total_ctc, 0))      AS ctc
        FROM `tabHiring Plan`
        WHERE {cond}
        GROUP BY budget_unit
        """,
        vals, as_dict=True,
    )
    by_unit = {r["unit"]: r for r in rows}
    regions = {}
    for unit, r in by_unit.items():
        regions.setdefault(UNIT_REGION.get(unit, ""), []).append((unit, r))

    out = []
    grand = {"hc": 0, "salary": 0, "ctc": 0}
    for reg in sorted(regions, key=lambda x: REGION_ORDER.get(x, 9)):
        sub = {"hc": 0, "salary": 0, "ctc": 0}
        urows = []
        for unit, r in sorted(regions[reg], key=lambda t: UNIT_ORDER.get(t[0], 9)):
            urows.append({"region": reg, "unit": unit, "kind": "unit",
                          "hc": int(r["hc"]), "salary": round(r["salary"] or 0), "ctc": round(r["ctc"] or 0)})
            sub["hc"] += r["hc"] or 0
            sub["salary"] += r["salary"] or 0
            sub["ctc"] += r["ctc"] or 0
        out.extend(urows)
        out.append({"region": reg, "unit": reg + " subtotal", "kind": "subtotal",
                    "hc": int(sub["hc"]), "salary": round(sub["salary"]), "ctc": round(sub["ctc"])})
        for k in grand:
            grand[k] += sub[k]
    out.append({"region": "", "unit": "TOTAL", "kind": "grand",
                "hc": int(grand["hc"]), "salary": round(grand["salary"]), "ctc": round(grand["ctc"])})
    return {"rows": out, "total_ctc": round(grand["ctc"]), "total_salary": round(grand["salary"])}


def _ramp(company, base_salary):
    cond = "status = 'Open'"
    vals = {}
    if company:
        cond += " AND company = %(company)s"
        vals["company"] = company
    rows = frappe.db.sql(
        f"""
        SELECT budget_unit AS unit, planned_month AS m,
               SUM(COALESCE(monthly_salary, 0)) AS sal
        FROM `tabHiring Plan`
        WHERE {cond}
        GROUP BY budget_unit, planned_month
        """,
        vals, as_dict=True,
    )
    add = {}
    for r in rows:
        if r["m"] in MONTHS:
            add.setdefault(r["unit"], {})[r["m"]] = r["sal"] or 0

    unit_series = []
    for unit in sorted(add, key=lambda u: UNIT_ORDER.get(u, 9)):
        cum, vals_out = 0, []
        for mo in MONTHS:
            cum += add[unit].get(mo, 0)
            vals_out.append(round(cum))
        unit_series.append({"unit": unit, "vals": vals_out})

    total, pct = [], []
    for i in range(6):
        added = sum(u["vals"][i] for u in unit_series)
        total.append(round(base_salary + added))
        pct.append(round((added / base_salary * 100), 1) if base_salary else 0)

    return {"months": MONTH_LBL, "current": [round(base_salary)] * 6,
            "units": unit_series, "total": total, "pct": pct}


def _designations(company):
    """Head count and cost per designation inside each budget unit."""
    cond, vals = _cond(company)
    rows = frappe.db.sql(
        f"""
        SELECT COALESCE(NULLIF(custom_budget_unit, ''), '(untagged)') AS unit,
               COALESCE(NULLIF(designation, ''), '(none)')            AS designation,
               COUNT(*)                                               AS hc,
               SUM(COALESCE(custom_total_salary, 0))                  AS salary,
               SUM(COALESCE(custom_monthly_ctc, 0))                   AS ctc
        FROM `tabEmployee`
        WHERE {cond}
          AND name NOT IN %(owners)s
        GROUP BY unit, designation
        ORDER BY unit, hc DESC
        """,
        dict(vals, owners=tuple(OWNERS)), as_dict=True,
    )

    out = {}
    for r in rows:
        out.setdefault(r["unit"], []).append({
            "designation": r["designation"],
            "hc": int(r["hc"] or 0),
            "salary": round(r["salary"] or 0),
            "ctc": round(r["ctc"] or 0),
        })

    # mirror the CEO split so designation rows still add up to the unit row
    if not company:
        ceo = frappe.db.get_value(
            "Employee", CEO,
            ["custom_budget_unit", "designation", "custom_total_salary", "custom_monthly_ctc"],
            as_dict=True,
        )
        if ceo and ceo.custom_budget_unit in out:
            share_sal = (ceo.custom_total_salary or 0) * CEO_DUBAI_SHARE
            share_ctc = (ceo.custom_monthly_ctc or 0) * CEO_DUBAI_SHARE
            label = ceo.designation or "(none)"
            for row in out[ceo.custom_budget_unit]:
                if row["designation"] == label:
                    row["salary"] = round(row["salary"] - share_sal)
                    row["ctc"] = round(row["ctc"] - share_ctc)
                    break
            out.setdefault("Dubai Office", []).append({
                "designation": CEO_ALLOC_LABEL,
                "hc": 0,
                "salary": round(share_sal),
                "ctc": round(share_ctc),
            })

    return out


def _months_back(n=6):
    """Last n months, oldest first, as (year, month)."""
    from datetime import date

    today = date.today()
    out = []
    for i in range(n - 1, -1, -1):
        mm, yy = today.month - i, today.year
        while mm <= 0:
            mm += 12
            yy -= 1
        out.append((yy, mm))
    return out


def _movement(company, current_ctc):
    """Joiners and leavers over the last 6 months and the effect on monthly CTC."""
    months = _months_back(6)
    start = "{0}-{1:02d}-01".format(months[0][0], months[0][1])

    vals = {"start": start, "owners": tuple(OWNERS)}
    ccond = ""
    if company:
        ccond = " AND company = %(company)s"
        vals["company"] = company

    joiners = frappe.db.sql(
        f"""
        SELECT DATE_FORMAT(date_of_joining, '%%Y-%%m') AS m,
               COUNT(*)                             AS hc,
               SUM(COALESCE(custom_monthly_ctc, 0)) AS ctc
        FROM `tabEmployee`
        WHERE date_of_joining IS NOT NULL
          AND date_of_joining >= %(start)s
          AND name NOT IN %(owners)s
          {ccond}
        GROUP BY m
        """,
        vals, as_dict=True,
    )

    leavers = frappe.db.sql(
        f"""
        SELECT DATE_FORMAT(relieving_date, '%%Y-%%m') AS m,
               COUNT(*)                             AS hc,
               SUM(COALESCE(custom_monthly_ctc, 0)) AS ctc
        FROM `tabEmployee`
        WHERE relieving_date IS NOT NULL
          AND relieving_date >= %(start)s
          AND name NOT IN %(owners)s
          {ccond}
        GROUP BY m
        """,
        vals, as_dict=True,
    )

    jmap = {r["m"]: r for r in joiners}
    lmap = {r["m"]: r for r in leavers}

    keys = ["{0}-{1:02d}".format(y, m) for y, m in months]
    labels = [MONTH_NAMES[m - 1] for _, m in months]

    total_added = sum(float((jmap.get(k) or {}).get("ctc") or 0) for k in keys)
    total_removed = sum(float((lmap.get(k) or {}).get("ctc") or 0) for k in keys)

    running = (current_ctc or 0) - total_added + total_removed

    rows = []
    n_joined = n_left = 0
    for key, label in zip(keys, labels):
        j = jmap.get(key) or {}
        l = lmap.get(key) or {}
        added = float(j.get("ctc") or 0)
        removed = float(l.get("ctc") or 0)
        closing = running + added - removed
        rows.append({
            "month": label,
            "key": key,
            "opening": round(running),
            "joined": int(j.get("hc") or 0),
            "added": round(added),
            "left": int(l.get("hc") or 0),
            "removed": round(removed),
            "closing": round(closing),
        })
        n_joined += int(j.get("hc") or 0)
        n_left += int(l.get("hc") or 0)
        running = closing

    ucond = ""
    uvals = {}
    if company:
        ucond = " AND company = %(company)s"
        uvals["company"] = company

    undated = frappe.db.sql(
        f"""
        SELECT COUNT(*) FROM `tabEmployee`
        WHERE status = 'Left'
          AND (relieving_date IS NULL OR relieving_date = '')
          {ucond}
        """,
        uvals,
    )
    undated_count = int(undated[0][0]) if undated else 0

    return {
        "rows": rows,
        "net_joined": n_joined,
        "net_added": round(total_added),
        "net_left": n_left,
        "net_removed": round(total_removed),
        "closing": round(running),
        "undated_leavers": undated_count,
    }

File: tripod_hr/tripod_hr/page/hr_budget_dashboard/hr_budget_dashboard.js
frappe.pages['hr-budget-dashboard'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'HR Budget Dashboard',
		single_column: true
	});

	var company_field = page.add_field({
		fieldname: 'company',
		label: 'Company',
		fieldtype: 'Select',
		options: [
			{ label: 'All (Tripod Media + Global)', value: 'All' },
			{ label: 'Tripod Media', value: 'Tripod Media FZ LLC' },
			{ label: 'Tripod Global', value: 'TRIPOD GLOBAL SHOPFIT MANUFACTURING COMPANY' }
		],
		default: 'All',
		change: function () { load(); }
	});

	page.add_inner_button('Refresh', function () { load(); });

	$('<style>').text(
		'.hbd{padding:22px 26px;max-width:1080px;margin:0 auto;color:#16233A;font-family:Inter,system-ui,sans-serif}' +
		'.hbd .kpis{display:grid;grid-template-columns:repeat(5,1fr);gap:12px;margin-bottom:8px}' +
		'.hbd .kpi{background:#fff;border:1px solid #E5E3DC;border-radius:12px;padding:14px 16px}' +
		'.hbd .kpi .l{font-size:12px;color:#6B7280;font-weight:500}' +
		'.hbd .kpi .v{font-size:26px;font-weight:600;line-height:1.1;margin-top:5px}' +
		'.hbd .kpi .d{font-size:12px;margin-top:3px;font-weight:600;color:#B8863B}' +
		'.hbd h2{font-size:18px;font-weight:600;margin:30px 0 4px}' +
		'.hbd .note{color:#6B7280;font-size:12px;margin:0 0 12px}' +
		'.hbd .card{background:#fff;border:1px solid #E5E3DC;border-radius:12px;overflow:hidden}' +
		'.hbd table{width:100%;border-collapse:collapse;font-size:12.5px}' +
		'.hbd th,.hbd td{padding:8px 12px;text-align:right;white-space:nowrap}' +
		'.hbd th:first-child,.hbd td:first-child{text-align:left}' +
		'.hbd thead th{background:#16233A;color:#fff;font-weight:500;font-size:11px;text-transform:uppercase;letter-spacing:.03em}' +
		'.hbd tbody td{border-bottom:1px solid #F0EEE8}' +
		'.hbd tr.sub td{background:#F5EBD6;font-weight:600}' +
		'.hbd tr.untag td{background:#FCEBEB;color:#A32D2D;font-weight:600}' +
		'.hbd tr.grand td{background:#16233A;color:#fff;font-weight:600;font-size:13px}' +
		'.hbd .pill{display:inline-block;font-size:10.5px;font-weight:600;padding:1px 8px;border-radius:20px}' +
		'.hbd .u{background:#E5EEF6;color:#2A5C8A}.hbd .k{background:#E2F1EA;color:#1B7A5A}' +
		'.hbd tr.dsg td{background:#FBFCFE;color:#3A4A63;font-size:12px;padding:6px 12px;border-bottom:1px solid #F5F4EF}' +
		'.hbd tr.dsg td:first-child{padding-left:30px}' +
		'.hbd .hclink{color:#185FA5;border-bottom:1px solid #185FA5;cursor:pointer}' +
		'.hbd .mvlink{color:#185FA5;border-bottom:1px solid #185FA5;cursor:pointer}' +
		'.hbd tr.mvsub td{background:#FBFCFE;font-size:11.5px;padding:5px 12px;border-bottom:1px solid #F5F4EF}' +
		'.hbd .chev{color:#9AA3B0;font-size:10px;margin-right:6px;cursor:pointer}' +
		'.hbd td.pos{color:#0F6E56}.hbd td.neg{color:#A32D2D}' +
		'.hbd .loading{padding:40px;text-align:center;color:#6B7280}'
	).appendTo('head');

	var $body = $('<div class="hbd"><div class="loading">Loading…</div></div>').appendTo(page.body);

	function fmt(n) { return Math.round(n || 0).toLocaleString(); }
	function regPill(reg) {
		if (reg === 'UAE') return '<span class="pill u">UAE</span>';
		if (reg === 'KSA') return '<span class="pill k">KSA</span>';
		return '';
	}

	function activeTable(a, dsg) {
		var h = '<div class="card"><table><thead><tr>' +
			'<th>Unit</th><th>Head count</th><th>Salary / mo</th><th>CTC / mo</th>' +
			'<th>Salary / yr</th><th>CTC / yr</th><th>Region</th></tr></thead><tbody>';
		var tot = a.current_ctc || 0;
		a.rows.forEach(function (r) {
			var cls = r.kind === 'subtotal' ? 'sub' : (r.kind === 'grand' ? 'grand' : (r.kind === 'untagged' ? 'untag' : ''));
			var pctv = tot ? (r.ctc / tot * 100).toFixed(1) + '%' : '';
			var kids = (r.kind === 'unit' || r.kind === 'untagged') ? (dsg && dsg[r.unit]) : null;
			var hasKids = kids && kids.length;

			var label = r.unit;
			var hcCell = fmt(r.hc);
			if (hasKids) {
				label = '<span class="chev" data-unit="' + r.unit + '">&#9654;</span>' + r.unit;
				hcCell = '<span class="hclink" data-unit="' + r.unit + '">' + fmt(r.hc) + '</span>';
			}

			h += '<tr class="' + cls + '"><td>' + label + '</td><td>' + hcCell + '</td><td>' +
				fmt(r.salary) + '</td><td>' + fmt(r.ctc) + '</td><td>' + fmt(r.salary_yr) + '</td><td>' +
				fmt(r.ctc_yr) + '</td><td>' + (r.kind === 'unit' ? regPill(r.region) : '') + '</td></tr>';

			if (hasKids) {
				kids.forEach(function (k) {
					h += '<tr class="dsg" data-parent="' + r.unit + '" style="display:none;">' +
						'<td>' + k.designation + '</td><td>' + (k.hc ? fmt(k.hc) : '–') + '</td><td>' +
						fmt(k.salary) + '</td><td>' + fmt(k.ctc) + '</td><td>' +
						fmt(k.salary * 12) + '</td><td>' + fmt(k.ctc * 12) + '</td><td></td></tr>';
				});
			}
		});
		return h + '</tbody></table></div>';
	}

	function planTable(p, curSalary) {
		var h = '<div class="card"><table><thead><tr>' +
			'<th>Unit</th><th>Head count</th><th>Salary / mo</th><th>CTC / mo</th><th>% incr</th><th>Region</th></tr></thead><tbody>';
		p.rows.forEach(function (r) {
			var cls = r.kind === 'subtotal' ? 'sub' : (r.kind === 'grand' ? 'grand' : '');
			var pctv = curSalary ? (r.salary / curSalary * 100).toFixed(1) + '%' : '';
			h += '<tr class="' + cls + '"><td>' + r.unit + '</td><td>' + fmt(r.hc) + '</td><td>' +
				fmt(r.salary) + '</td><td>' + fmt(r.ctc) + '</td><td>' + pctv + '</td><td>' +
				(r.kind === 'unit' ? regPill(r.region) : '') + '</td></tr>';
		});
		return h + '</tbody></table></div>';
	}

	function rampTable(rp) {
		var h = '<div class="card"><table><thead><tr><th>Line</th>';
		rp.months.forEach(function (m) { h += '<th>' + m + '</th>'; });
		h += '</tr></thead><tbody>';
		h += '<tr><td>Current staff</td>';
		rp.current.forEach(function (v) { h += '<td>' + fmt(v) + '</td>'; });
		h += '</tr>';
		rp.units.forEach(function (u) {
			h += '<tr><td>' + u.unit + '</td>';
			u.vals.forEach(function (v) { h += '<td>' + (v ? fmt(v) : '–') + '</td>'; });
			h += '</tr>';
		});
		h += '<tr class="sub"><td>Total payout</td>';
		rp.total.forEach(function (v) { h += '<td>' + fmt(v) + '</td>'; });
		h += '</tr><tr class="grand"><td>% increase</td>';
		rp.pct.forEach(function (v) { h += '<td>' + v + '%</td>'; });
		h += '</tr></tbody></table></div>';
		return h;
	}

	function movementTable(m) {
		var h = '<div class="card"><table><thead><tr>' +
			'<th>Month</th><th>Opening CTC</th><th>Joined</th><th>Added</th>' +
			'<th>Left</th><th>Removed</th><th>Closing CTC</th></tr></thead><tbody>';

		m.rows.forEach(function (r, i) {
			var jCell = r.joined
				? '<span class="mvlink" data-row="' + i + '" data-kind="j">' + fmt(r.joined) + '</span>'
				: '–';
			var lCell = r.left
				? '<span class="mvlink" data-row="' + i + '" data-kind="l">' + fmt(r.left) + '</span>'
				: '–';
			h += '<tr><td>' + r.month + '</td><td>' + fmt(r.opening) + '</td>' +
				'<td>' + jCell + '</td>' +
				'<td class="pos">' + (r.added ? '+' + fmt(r.added) : '–') + '</td>' +
				'<td>' + lCell + '</td>' +
				'<td class="neg">' + (r.removed ? '\u2212' + fmt(r.removed) : '–') + '</td>' +
				'<td>' + fmt(r.closing) + '</td></tr>';

			// hidden breakdown rows for joined
			if (r.joined) {
				h += '<tr class="mvsub" data-parent="' + i + '-j" style="display:none;">' +
					'<td></td><td style="text-align:right;color:#3A4A63;">Office Staff</td>' +
					'<td>' + fmt(r.joined_office) + '</td><td colspan="4"></td></tr>';
				h += '<tr class="mvsub" data-parent="' + i + '-j" style="display:none;">' +
					'<td></td><td style="text-align:right;color:#3A4A63;">Labour</td>' +
					'<td>' + fmt(r.joined_labour) + '</td><td colspan="4"></td></tr>';
			}
			// hidden breakdown rows for left
			if (r.left) {
				h += '<tr class="mvsub" data-parent="' + i + '-l" style="display:none;">' +
					'<td></td><td colspan="3"></td><td style="text-align:left;color:#3A4A63;">Office Staff ' + fmt(r.left_office) + '</td><td colspan="2"></td></tr>';
				h += '<tr class="mvsub" data-parent="' + i + '-l" style="display:none;">' +
					'<td></td><td colspan="3"></td><td style="text-align:left;color:#3A4A63;">Labour ' + fmt(r.left_labour) + '</td><td colspan="2"></td></tr>';
			}
		});

		h += '<tr class="grand"><td>NET</td><td></td><td>' + fmt(m.net_joined) + '</td><td>+' +
			fmt(m.net_added) + '</td><td>' + fmt(m.net_left) + '</td><td>\u2212' +
			fmt(m.net_removed) + '</td><td>' + fmt(m.closing) + '</td></tr>';

		h += '</tbody></table></div>';

		if (m.undated_leavers) {
			h += '<p class="note" style="color:#A32D2D;margin-top:8px;">' + m.undated_leavers +
				' employee(s) marked Left have no relieving date, so they are not in the Left column. ' +
				'Add the date on their Employee record to include them.</p>';
		}
		return h;
	}

	function render(d) {
		var a = d.active, p = d.plan, rp = d.ramp;
		var dsg = d.designations || {};
		var mv = d.movement;
		var projected = (a.current_ctc || 0) + (p.total_ctc || 0);
		var pct = a.current_ctc ? (p.total_ctc / a.current_ctc * 100).toFixed(1) : 0;
		var grand = a.rows.filter(function (r) { return r.kind === 'grand'; })[0] || { hc: 0 };

		var html = '<div class="kpis">' +
			kpi('Active headcount', fmt(grand.hc), '') +
			kpi('Monthly Salary', fmt(a.current_salary), '') +
			kpi('Monthly CTC', fmt(a.current_ctc), '') +
			kpi('Planned hires', fmt(p.rows.filter(function (r) { return r.kind === 'grand'; })[0].hc), '') +
			kpi('Projected CTC', fmt(projected), '+' + pct + '%') +
			'</div>';

		html += '<h2>1 · Active staff</h2><p class="note">Live from Employee master (status = Active). Click a head count to see its designations.</p>' + activeTable(a, dsg);
		// Sections 2 & 3 hidden for now (kept in code, can be restored):
		// html += '<h2>2 · Budget increase — Jul–Dec 2026</h2><p class="note">Open positions in the Hiring Plan. % = share of current payroll added.</p>' + planTable(p, a.current_salary);
		// html += '<h2>3 · Likely monthly payable — Jul–Dec</h2><p class="note">Current staff flat; open positions phase in by planned month (salary, cumulative).</p>' + rampTable(rp);

		if (mv && mv.rows && mv.rows.length) {
			html += '<h2>4 · Movement — joiners &amp; leavers</h2>' +
				'<p class="note">Last 6 months. Who joined, who left, and the effect on monthly CTC.</p>' +
				movementTable(mv);
		}

		$body.html(html);

		$body.find('.mvlink').on('click', function () {
			var key = $(this).attr('data-row') + '-' + $(this).attr('data-kind');
			var $subs = $body.find('tr.mvsub').filter(function () {
				return $(this).attr('data-parent') === key;
			});
			var showing = $subs.first().is(':visible');
			$subs.toggle(!showing);
		});

		$body.find('.hclink, .chev').on('click', function () {
			var unit = $(this).attr('data-unit');
			var $kids = $body.find('tr.dsg').filter(function () {
				return $(this).attr('data-parent') === unit;
			});
			var showing = $kids.first().is(':visible');
			$kids.toggle(!showing);
			$body.find('.chev').filter(function () {
				return $(this).attr('data-unit') === unit;
			}).html(showing ? '&#9654;' : '&#9660;');
		});
	}

	function kpi(l, v, d) {
		return '<div class="kpi"><div class="l">' + l + '</div><div class="v">' + v + '</div>' +
			(d ? '<div class="d">' + d + '</div>' : '<div class="d" style="color:#6B7280">&nbsp;</div>') + '</div>';
	}

	function load() {
		$body.html('<div class="loading">Loading…</div>');
		var c = company_field.get_value();
		frappe.call({
			method: 'tripod_hr.tripod_hr.page.hr_budget_dashboard.hr_budget_dashboard.get_budget_dashboard',
			args: { company: (c && c !== 'All') ? c : null },
			callback: function (r) {
				if (r.message) { render(r.message); }
				else { $body.html('<div class="loading">No data.</div>'); }
			}
		});
	}

	load();
};
